### ISSUE TITLE
Resolving shapes of outputs of Linalg operation on tensors

### ISSUE THREAD

**Issue by MaheshRavishankar** [google/iree#4205]
With Linalg operation, there are some cases where the bounds of the computation might be dependent on the output.
For example consider the following broadcast operation in Linalg on buffers

```
linalg.generic {indexing_maps=[affine_map<(d0, d1) -> (d1)>,
                               affine_map<(d0, d1) -> (d0, d1)>],
                iterator_type = ["parallel", "parallel"]}
               ins(%0 : memref<4xf32>)
               outs(%1 : memref<?x4xf32>) {
  ^bb0(%arg0 : f32) : 
    linalg.yield %arg0
  }
```

Here the loop bounds for the `d0` dimension of the nested loop represented by Linalg operation is the shape of the output
```
%c0 = constant 0 : index
dim %1, %c0 : memref<?x4xf32>
```

In buffer world the memref is part of the operands of the linalg operation. So there is no problem with the semantics here. In tensor world the same operation looks like this:

```
%1 = linalg.generic {indexing_maps=[affine_map<(d0, d1) -> (d1)>,
                                    affine_map<(d0, d1) -> (d0, d1)>],
                     iterator_type = ["parallel", "parallel"]}
                    ins(%0 : tensor<4xf32>) {
  ^bb0(%arg0 : f32) : 
    linalg.yield %arg0
  } -> tensor<?x4xf32>
```

This means that the bounds of the computation now depends on the tensor being created. In effect, the operation itself "might" be considered ill-formed. Concretely, there might be cases where a linalg operation on tensors is dead except for its use in `dim` statements of bounds computation. 

THere are several ways to address this issue

**Method 1:**

Assume that the shape dialect (`shapex.*`) provides the missing information w.r.t shapes. In reality, this is not a linalg specific problem. Any operation that produces tensors has this issue w.r.t to dynamic shapes. Seeing the above as an instance of that problem, we can use the `shapex` dialect as follows

```
%1 = linalg.generic {indexing_maps=[affine_map<(d0, d1) -> (d1)>,
                                    affine_map<(d0, d1) -> (d0, d1)>],
                     iterator_type = ["parallel", "parallel"]}
                    ins(%0 : tensor<4xf32>) {
  ^bb0(%arg0 : f32) : 
    linalg.yield %arg0
  } -> tensor<?x4xf32>
%2 = !shapex.make_ranked_shape %s : (index) -> !shapex.ranked_shape<[?,4]>
%3 = !shapex.tie_shape %1, %2 : tensor<?x4xf32>, !shapex.ranked_shape<[?,4]>
```

Then it will be upto the shape resolution pass to decouple the `linalg.generic` and the shape of the result tensor. IIUC this is pretty much what the shape dialect is meant to be doing.

**Method 2:**

The issue only arises for `linalg` operations on tensors that have only `parallel` iterator types. When a linalg operation on tensors has a `reduction` iterator type, the operation has an additional `init` tensor which is of the same shape of the output. The semantics of linalg operation can be changed to make all operations have `init` tensors. In that case the above operation would be

```
%1 = linalg.generic {indexing_maps=[affine_map<(d0, d1) -> (d1)>,
                                    affine_map<(d0, d1) -> (d0, d1)>],
                     iterator_type = ["parallel", "parallel"]}
                    ins(%0 : tensor<4xf32>)
                    init(%i : tensor<?x4xf32>) {
  ^bb0(%arg0 : f32) : 
    linalg.yield %arg0
  } -> tensor<?x4xf32>
```

Then you can use the shape of the `init` tensor to compute the loop bounds. This semantics implies that `init` tensor operand is only used for the shape. It also means that you might have to materialize tensor values just for the shape

```
%i = linalg.init_tensor %s : tensor<?x4xf32>
%1 = linalg.generic {indexing_maps=[affine_map<(d0, d1) -> (d1)>,
                                    affine_map<(d0, d1) -> (d0, d1)>],
                     iterator_type = ["parallel", "parallel"]}
                    ins(%0 : tensor<4xf32>)
                    init(%i : tensor<?x4xf32>) {
  ^bb0(%arg0 : f32) : 
    linalg.yield %arg0
  } -> tensor<?x4xf32>
```
It would be upto the bufferization pass to then "fold" away the `%i` and not materialize two separate buffers, one for the output of the result of `linalg.init_tensor` and one for the output of `linalg.generic`. This approach is effectively the Method 1 in reverse. `linalg.init_tensor` is effectively doing the role of a `!shapex.tie_shape` but by actually materializing a tensor.

**Method 3**

The linalg operation itself can be extended to contain shape information for the output which in most cases is redundant information

```
%1 = linalg.generic {indexing_maps=[affine_map<(d0, d1) -> (d1)>,
                                    affine_map<(d0, d1) -> (d0, d1)>],
                     iterator_type = ["parallel", "parallel"]}
                    ins(%0 : tensor<4xf32>)
                    shape[%s, 4] {
  ^bb0(%arg0 : f32) : 
    linalg.yield %arg0
  } -> tensor<?x4xf32>
```

The `shape` operands tell you the shape of the output. For the most cases, the entries of the `shape` can be `?` since the bounds of the computation doesn't depend on the output shape. But for cases above, they need to be specified or else the operation would be ill-formed. This gets fairly non-trivial as to what are the cases where the `shape` cannot be `?`. The problem with this though is if you tile the operation for the case where one or more of the `shape` entries are not `?`, then you need to compute the corresponding value for the tiled operation. On the other hand, tHis is already done in the tiling transformation of linalg operations.

On balance I think method 3 is superior, but requires change to Linalg Structured op semantics. In the meantime at least for IREE, Method 1 works.

**Comment by benvanik** [comment]
Cool - this looks like it comes prior to what I'm doing now with outlining in #4140. The assumption is that once a `flow.dispatch.workgroups` op is formed there are appropriate shape ties in the external IR:

```mlir
func @dynamicShapeDispatch(%arg0 : tensor<7x?x24x?xf32>) -> tensor<?x?x1024xf32> {
  %c1 = constant 1 : index
  %c3 = constant 3 : index
  %dim1 = dim %arg0, %c1 : tensor<7x?x24x?xf32>
  %dim3 = dim %arg0, %c3 : tensor<7x?x24x?xf32>
  %x = constant 1024 : index
  %y = constant 512 : index
  %shape0 = shapex.make_ranked_shape %dim1, %dim3 : (index, index) -> !shapex.ranked_shape<[7,?,24,?]>
  %arg0_shaped = shapex.tie_shape %arg0, %shape0 : tensor<7x?x24x?xf32>, !shapex.ranked_shape<[7,?,24,?]>
  %0 = flow.dispatch.workgroups[%x, %y](%arg0_shaped) : (tensor<7x?x24x?xf32>) -> tensor<?x?x1024xf32> = (
    %arg : !flow.dispatch.input<7x?x24x?xf32>, %ret : !flow.dispatch.output<?x?x1024xf32>
  ) {
    // ...
  }
  %1 = shapex.make_ranked_shape %c3, %c1 : (index, index) -> !shapex.ranked_shape<[?,?,1024]>
  %2 = shapex.tie_shape %0, %1 : tensor<?x?x1024xf32>, !shapex.ranked_shape<[?,?,1024]>
  return %2 : tensor<?x?x1024xf32>
}
```

Those are then used to propagate the shapes into the outlined dispatch region (example outlined contents doing shape stuff):

```mlir
  flow.executable @dynamicShapeDispatch_dispatch_0 attributes {sym_visibility = "private"} {
    flow.dispatch.entry @dynamicShapeDispatch_dispatch_0 attributes {signature = (tensor<7x?x24x?xf32>) -> tensor<?x?x1024xf32>, workgroup_rank = 2 : index}
    module  {
      func @dynamicShapeDispatch_dispatch_0(%arg0: !flow.dispatch.input<7x?x24x?xf32>, %arg1: !flow.dispatch.output<?x?x1024xf32>, %arg2: index, %arg3: index, %arg4: index, %arg5: index) {
        %0 = shapex.make_ranked_shape %arg2, %arg3 : (index, index) -> !shapex.ranked_shape<[7,?,24,?]>
        %1 = flow.dispatch.tie_shape %arg0, %0 : (!flow.dispatch.input<7x?x24x?xf32>, !shapex.ranked_shape<[7,?,24,?]>) -> !flow.dispatch.input<7x?x24x?xf32>
        %2 = shapex.make_ranked_shape %arg4, %arg5 : (index, index) -> !shapex.ranked_shape<[?,?,1024]>
        %3 = flow.dispatch.tie_shape %arg1, %2 : (!flow.dispatch.output<?x?x1024xf32>, !shapex.ranked_shape<[?,?,1024]>) -> !flow.dispatch.output<?x?x1024xf32>
        %workgroup_rank = flow.dispatch.workgroup.rank : index
        %shape = flow.dispatch.shape %1 : !flow.dispatch.input<7x?x24x?xf32> -> !shapex.ranked_shape<[7,?,24,?]>
        %4 = shapex.ranked_dim %shape[1] : !shapex.ranked_shape<[7,?,24,?]> -> index
        %5 = shapex.ranked_dim %shape[3] : !shapex.ranked_shape<[7,?,24,?]> -> index
        "test.sink_shape_arg"(%4, %5) : (index, index) -> ()
        %shape_0 = flow.dispatch.shape %3 : !flow.dispatch.output<?x?x1024xf32> -> !shapex.ranked_shape<[?,?,1024]>
        %6 = shapex.ranked_dim %shape_0[0] : !shapex.ranked_shape<[?,?,1024]> -> index
        %7 = shapex.ranked_dim %shape_0[1] : !shapex.ranked_shape<[?,?,1024]> -> index
        "test.sink_shape_ret"(%6, %7) : (index, index) -> ()
        %8 = flow.dispatch.input.load %1 : !flow.dispatch.input<7x?x24x?xf32> -> tensor<7x?x24x?xf32>
        %9 = shapex.get_ranked_shape %8 : tensor<7x?x24x?xf32> -> !shapex.ranked_shape<[7,?,24,?]>
        %10 = "test.tile_math"(%8, %9, %shape_0) : (tensor<7x?x24x?xf32>, !shapex.ranked_shape<[7,?,24,?]>, !shapex.ranked_shape<[?,?,1024]>) -> tensor<?x?x1024xf32>
        flow.dispatch.output.store %10, %3 : tensor<?x?x1024xf32> -> !flow.dispatch.output<?x?x1024xf32>
        return
      }
    }
  }
```

Which after canonicalization folds to about what you could expect:

```mlir
  flow.executable @dynamicShapeDispatch_dispatch_0 attributes {sym_visibility = "private"} {
    flow.dispatch.entry @dynamicShapeDispatch_dispatch_0 attributes {signature = (tensor<7x?x24x?xf32>) -> tensor<?x?x1024xf32>, workgroup_rank = 2 : index}
    module  {
      func @dynamicShapeDispatch_dispatch_0(%arg0: !flow.dispatch.input<7x?x24x?xf32>, %arg1: !flow.dispatch.output<?x?x1024xf32>, %arg2: index, %arg3: index, %arg4: index, %arg5: index) {
        %0 = shapex.make_ranked_shape %arg2, %arg3 : (index, index) -> !shapex.ranked_shape<[7,?,24,?]>
        %1 = flow.dispatch.tie_shape %arg0, %0 : (!flow.dispatch.input<7x?x24x?xf32>, !shapex.ranked_shape<[7,?,24,?]>) -> !flow.dispatch.input<7x?x24x?xf32>
        %2 = shapex.make_ranked_shape %arg4, %arg5 : (index, index) -> !shapex.ranked_shape<[?,?,1024]>
        %3 = flow.dispatch.tie_shape %arg1, %2 : (!flow.dispatch.output<?x?x1024xf32>, !shapex.ranked_shape<[?,?,1024]>) -> !flow.dispatch.output<?x?x1024xf32>
        "test.sink_shape_arg"(%arg2, %arg3) : (index, index) -> ()
        "test.sink_shape_ret"(%arg4, %arg5) : (index, index) -> ()
        %4 = flow.dispatch.input.load %1 : !flow.dispatch.input<7x?x24x?xf32> -> tensor<7x?x24x?xf32>
        %5 = shapex.get_ranked_shape %4 : tensor<7x?x24x?xf32> -> !shapex.ranked_shape<[7,?,24,?]>
        %6 = "test.tile_math"(%4, %5, %2) : (tensor<7x?x24x?xf32>, !shapex.ranked_shape<[7,?,24,?]>, !shapex.ranked_shape<[?,?,1024]>) -> tensor<?x?x1024xf32>
        flow.dispatch.output.store %6, %3 : tensor<?x?x1024xf32> -> !flow.dispatch.output<?x?x1024xf32>
        return
      }
    }
  }
```

The big area this glosses over is how the flow.dispatch.input/output.load/store ops propagate shapes of tiles - I think that's where things intersects with this issue. There's enough metadata here to take %1 given to flow.dispatch.input.load and get the shape of the full thing, and then the output shape is based on the sizes of the load (omitted in this test). So as long as you can do that in linalg and materialize them during construction of the regions we should be able to make that flow through.

**Comment by silvasean** [comment]
I would recommend using the init tensor ("method 2") for now. Basically we just need an op that creates a tensor with the property `dim(op(x), 0) -> x`, and then we can DCE `op` after it becomes unused. In terms of making sure that the tensor never gets materialized, it's basically the same problem as when we sink constants into dispatch regions, except in this case as we sink it we need to carry a shape.

Note that `linalg.init_tensor` that you propose is basically the same as `tensor.undef` AFAICT, and has the same caveats. I'm happy to drive the discussion upstream for that, as I think I have a good design for that worked out. Stuffing it inside linalg seems kind of hacky (especially if it doesn't get the care and attention for review as tensor.undef would, despite being the same op); I would prefer to keep this IREE-specific for now `iree.undef_tensor %s0, %s1, sN` (where `%s_i` is the size of each dimension for a ranked output). It really belongs upstream as tensor.undef though.


Iota is another interesting one here (since it takes no tensor arguments logically). I think we can also address that with `tensor.undef` as an init tensor + indexed generic. We should keep an eye on that to see if that representation becomes unwieldy. We might want to look into method 3 if that's the case, but I'm very afraid of making the structured op interfaces more complex so would like to avoid that if possible.

**Comment by benvanik** [comment]
Init tensor is going to make it more complicated to form the dispatch regions (what mahesh will be doing, so if he says otherwise then I defer to him ;) and I'm not sure what it gains - would really like to avoid having additional magic that requires special handling in later passes (bufferization/etc). If we were going to code anything over what we have today with method 1 I'd rather go right for method 3 where things can remain nice and localized.

**Comment by hanhanW** [comment]
I have a prototype on lowering ReduceOp to Linalg on tensors based on *Method 2* -- with `linalg.init_tensor`. Hope this could get you more ideas what it would look like. I'm still exporting it to Github, but can CC you the internal change.

Regarding what @silvasean mentioned `iree.undef_tensor %s0, %s1, sN`, I just found that there is a similar thing in std dialect: `dynamic_tensor_from_elements`.

```
    This operation creates a dynamically sized tensor with elements of any type.
    It expects one index operand per dynamic extent of the result tensor.

    The body region defines the tensor's elements. It takes index operands as
    its region arguments that span the index space. The element at the given
    position is yielded with the `yield` operation (see `YieldOp`). There is
    no defined ordering to the invocations of the body. It is conceptually
    a "parallel map" operation.

    Example:

    ```mlir
      %tnsr = dynamic_tensor_from_elements %m, %n {
      ^bb0(%i : index, %j : index, %k : index):
        ...
        yield %elem : f32
      } : tensor<?x3x?f32>
    ```
```

I think the prototype can work with this op even `linalg.init_tensor` is not introduced. So we could focus on the right things without maintaining `linalg.init_tensor`.

**Comment by silvasean** [comment]
@benvanik  method 2 and 3 are basically the same. The question is, from an IR design perspective, is there value to introducing the shape thing into the linalg ops themselves, or is it fairly orthogonal functionality that can just live in a separate op. It's not obvious to me that we need linalg ops themselves to know about this -- it kind of smells to me like having a conv op with a hardcoded set of fusions on the op itself.

@hanhanW `dynamic_tensor_from_elements` should have the properties we need (`dim(op(x), 0) -> x`, and then we can DCE op after it becomes unused). Good find!

**Comment by hanhanW** [comment]
The patch is out, but has some redundant changes: https://github.com/google/iree/pull/4206
For prototype on InitTensor op, please see reduce.mlir files, one is lowering to Linalg on tensors, and another is lowering to Linalg on buffers with whole pipeline.

**Comment by benvanik** [comment]
Not sure I follow - it's quite a bit worse coupling to have an init tensor than to have the shape of the output provided explicitly. The problem that causes all of this to come about is that we *don't* have a nice explicit way to specify shapes, and instead rely on non-local information that requires correct construction (and maintenance) of the IR. Introducing additional tightly-coupled structural ops like undefined tensors/etc just feels like hacks on top of hacks - all you really want is a way to know what the output shape is.

When it comes to forming the regions it's really important we don't have to perform significant work to preserve the shape invariants of the ops that are moved into the region (both on the inside and the outside) and any additional non-local walking of the IR we have to do is just additional opportunities for complexity. Think of what happens after things like the initialization tensors are moved out of the block containing the operations or passed across function call boundaries - let along passed through a CFG. One should be able to look at a linalg op and know everything required for that linalg op just by looking at it.

**Comment by benvanik** [comment]
(but again, if mahesh thinks he can tile and fuse with init tensors *and* the init tensors end up being eliminated entirely I don't really care - just can't have them being pulled across the dispatch region boundary)

**Comment by silvasean** [comment]
The thing to observe is that the init tensor is needed anyway when reductions are present, and are actually needed even in the absence of reductions when doing tiling for linalg on tensors (I forget why; this is coming from Nicolas). So if we need to have the init tensor already for many cases (and the code to handle it), then there doesn't seem to be much of a point in adding any additional thing.

For example, we certainly are going to want to be able to do the right thing for `broadcast %scalar to %shape` or other situations where the init tensor is an easy to materialize constant. `tensor.undef %shape` will use that exact same code path.

**Comment by MaheshRavishankar** [comment]
> I would recommend using the init tensor ("method 2") for now. Basically we just need an op that creates a tensor with the property `dim(op(x), 0) -> x`, and then we can DCE `op` after it becomes unused. In terms of making sure that the tensor never gets materialized, it's basically the same problem as when we sink constants into dispatch regions, except in this case as we sink it we need to carry a shape (and I think that with Ben's new flow stuff we might avoid this problem entirely; still parsing that design though).
> 
> Note that `linalg.init_tensor` that you propose is basically the same as `tensor.undef` AFAICT, and has the same caveats. I'm happy to drive the discussion upstream for that, as I think I have a good design for that worked out. Stuffing it inside linalg seems kind of hacky (especially if it doesn't get the care and attention for review as tensor.undef would, despite being the same op); I would prefer to keep this IREE-specific for now `iree.undef_tensor %s0, %s1, sN` (where `%s_i` is the size of each dimension for a ranked output). It really belongs upstream as tensor.undef though.
> 
> Iota is another interesting one here (since it takes no tensor arguments logically). I think we can also address that with `tensor.undef` as an init tensor + indexed generic. We should keep an eye on that to see if that representation becomes unwieldy. We might want to look into method 3 if that's the case, but I'm very afraid of making the structured op interfaces more complex so would like to avoid that if possible.

This all makes sense. In https://reviews.llvm.org/D93077 I added a `linalg.init_tensor` but really I need `tensor.undef`. I also added a fake initialization value to side-step `undef` but I really dont want to do that either. So `tensor.undef` is fine.
In my view though `init` tensor approach is the least satisfactory of all. I see two issues with it
1) In Linalg the init tensor was used with operations with reduction semantics. So the actual value in the tensor is meaningful. For cases without reduction semantics the value is meaningless since it is all over written anyway which begs the question why even have that tensor. What you really need is the shape and solving that problem would be better.
2) You are also now putting weight on a bufferization pass to fold these away which is mixing concerns. We are adding more dialect specific concerns to a generic pass that is converting buffers to memref. I am not aware of all the details of bufferization pass but to me this smells like a leak.

Method 1 and 3 dont need bufferization to do anything special. Method 1 still relies on the shape dialect to resolve shapes but that is what it is meant to do, so its not adding any "linalg" specific constraints to it.

**Comment by benvanik** [comment]
Ah --- in the new world init tensors aren't needed and are explicitly designed out from everything after (at the latest) the tiling stage - we can go right from linalg ops to tiled representations. Relates to the big issue tracking the unneeded memsets/loads that get introduced by the init tensors naive lowering today.

**Comment by MaheshRavishankar** [comment]
> Ah --- in the new world init tensors aren't needed and are explicitly designed out from everything after (at the latest) the tiling stage - we can go right from linalg ops to tiled representations. Relates to the big issue tracking the unneeded memsets/loads that get introduced by the init tensors naive lowering today.

Not necessarily true, but thats a separate discussion

**Comment by MaheshRavishankar** [comment]
> but again, if mahesh thinks he can tile and fuse with init tensors _and_ the init tensors end up being eliminated entirely I don't really care - just can't have them being pulled across the dispatch region boundary

tile + fuse works with init tensors (and works without it too. See patch https://reviews.llvm.org/D93077 ). The `init_tensor`s can be eliminated at bufferization time, but as I said above this seems to be an undue burden on the bufferization. The way I see it, if I need to replace bufferization pass entirely what are the things the new pass needs to preserve. The lesser the things here, the better and more modular it is.

**Comment by silvasean** [comment]
> Method 1 and 3 dont need bufferization to do anything special. Method 1 still relies on the shape dialect to resolve shapes but that is what it is meant to do, so its not adding any "linalg" specific constraints to it.

I don't understand how method 2 this is asking bufferization to do something special. We literally just need to apply one pattern to bypass the dim ops and then DCE the tensor.undef. This all happens before we even convert to buffers. Everything else "just works" with all the existing handling of init tensors (which we need anyway due to ops with reduction semantics).

**Comment by MaheshRavishankar** [comment]
Ok, good point about bufferization. I didnt think of that. But the main issue is that now **all** ops need `init` tensor which seems too heavy weight when the values of the tensor are not really needed. Having an entire tensor for just its shape seems off

**Comment by silvasean** [comment]
It seems way easier for all ops to have `init` tensors than for something like method 3 which makes linalg ops way more complex.

Method 1 doesn't make sense to me -- the whole point of what we are discussing here is how linalg should model ops with tensor outputs whose size cannot be determined as a function of the `ins`. The `shapex` ops are essentialy an annotation which are part of the IREE implementation that are generated based on an op's own knowledge of its output shapes -- so it's a different (lower) layer and cannot factor into a solution to the core problem: how to have linalg ops whose output size is not a function of the `ins`.

**Comment by MaheshRavishankar** [comment]
W.R.T Method 1, the assumption was that whatever lowers into IREE has that information. So this is an assumption that as specified Linalg on tensors does not have all the information about the op by itself and that something that lowers into IREE does and it embeds that information into IR using the shape dialect. I do agree it is a stop-gap + lets make progress while we fix the issue for good kind of solution.

Between Method 2 and Method 3, I guess we have some disagreement on what is the "cleaner" solution. Lets see whats the consensus is. We can make either work

**Comment by MaheshRavishankar** [comment]
@hanhanW I missed the comment about `dynamic_tensor_from_elements` and yes that is what we need. It does work for this particular case and Method 2. That is a good find indeed. Lets discuss this tomorrow.

**Comment by nicolasvasilache** [comment]
+1 on `@benvanik method 2 and 3 are basically the same.` Will catch up with the rest a bit later, please come to the 9am PST meeting :)

**Comment by nicolasvasilache** [comment]
Added some thoughts in https://llvm.discourse.group/t/linalg-and-shapes/2421 that also capture some of the discussion with @pifon2a and @sherhut. This also explains why we are in this situation, from the linalg perspective. The hardest thing for me was to let go of the illusion that `linalg.generic` has all it needs for range inference: it almost and we can often drop many operands but in practice consistency and lack of surprise of OpInterface and op semantics significantly trump nice syntax for `linalg.generic`: this should not be its job.

Some replies to some of the comments. Time difference + low-BW medium gets you a wall of text :p 

> Cool - this looks like it comes prior to what I'm doing now with outlining in #4140. The assumption is that once a `flow.dispatch.workgroups` op is formed there are appropriate shape ties in the external IR:

Note: `!shapex.ranked_shape<[7,?,24,?]>`resembles `tensor<7x?x24x?xt, shape_only=1>`. We could also have a new type in core or an extra bit to tensor and memref type. Whichever solution we end up with, we want this type to "just work" with shape dialect ops, subview/subtensor/subtensor_insert, their canonicalizations and all linalg transformations. So we will need something with more static information than the all-erased `!shape.shape` type.

> I would recommend using the init tensor ("method 2") for now. Basically we just need an op that creates a tensor with the property dim(op(x), 0) -> x, and then we can DCE op after it becomes unused.

Yes, this is what "init_tensor form" is. The biggest problem I had to pull the trigger here is the amount of extra / redundant information involved. But after exhausting all other options I am left with: *"all linalg on tensor ops must always be in init_tensor/output shape form because that is the only way to keep things globally consistent". Turning output tensors into "shape only" information is an optimization that will be easy to do and propagate once the proper type exists.

> linalg.init_tensor that you propose is basically the same as tensor.undef AFAICT, and has the same caveats.

I don't think we want to condition anything on `tensor.undef` for now. We already know we need a proper `tensor.constant %v, %shape_info: tensor<?x8xf32>` to support reductions. Undef semantics can be dealt with later when we see a concrete need on the optimization side. From LLVM's langref: *"Undefined values are useful because they indicate to the compiler that the program is well defined no matter what value is used. This gives the compiler more freedom to optimize. "* I think we can safely keep this orthogonal for now and get the benefits later.

> If we were going to code anything over what we have today with method 1 I'd rather go right for method 3 where things can remain nice and localized.

The easy step forward here would be to have a `shape_only` bit on tensor and memref. In the absence of that, we will have quite more refactorings to connect everything: the key thing we want to "just work" is canonicalizations (`subview`/`subensor`/`subtensor_insert`) and transformations.

> Regarding what @silvasean mentioned iree.undef_tensor %s0, %s1, sN, I just found that there is a similar thing in std dialect: dynamic_tensor_from_elements. 

Cool find :) In a few weeks I expect it will look like `linalg.index_generic %shape { region } : tensor<?x8xf32, shape_only=1>` modulo what type we'll use for that shape information. We could also go directly for `linalg.fill %shape, %val : tensor<?x8xf32, shape_only=1>`. But let's start with `dynamic_tensor_from_elements` indeed.

>  Introducing additional tightly-coupled structural ops like undefined tensors/etc just feels like hacks on top of hacks - all you really want is a way to know what the output shape is.

My strong turn the lights on preference is `tensor<?x4xf32, shape_only=1>` (see below), but I'll work with whatever others want to implement (and that composes with all the stuff mentioned below) :) .

> and are actually needed even in the absence of reductions when doing tiling for linalg on tensors (I forget why; this is coming from Nicolas)

Tiling on tensors is done via destructive updates + yield values: there needs to be a tensor into which we write and yield. There are other potential designs but as we unroll them we invariably get to nested `linalg.generic` that have significantly deeper implications and are longer term. 

> In Linalg the init tensor was used with operations with reduction semantics. So the actual value in the tensor is meaningful. For cases without reduction semantics the value is meaningless since it is all over written anyway which begs the question why even have that tensor. What you really need is the shape and solving that problem would be better.

There is a lot of value starting with tensors because making things that "just work" with consistent structured ops semantics + canonicalizations + transformations is really the hard part. In the longer term, it would be whatever type we end up with that talks to the shape dialect and the `subview`/`subtensor` etc. 

> You are also now putting weight on a bufferization pass to fold these away which is mixing concerns. We are adding more dialect specific concerns to a generic pass that is converting buffers to memref. I am not aware of all the details of bufferization pass but to me this smells like a leak.

I view this as an optimization problem: in a first impl. we'll just allocate like crazy. Later, when we have the type we want and interfaces are defined properly we can just flip the type bit on the outputs that are just shape.
By far the easiest would be to directly start with an extra bit `memref<?x8xf32, shape_only=1>` and `tensor<4x?xf32>` because everything would just work. This will likely be contentious in core but is by far the easiest for incremental progress.

> Ah --- in the new world init tensors aren't needed and are explicitly designed out from everything after (at the latest) the tiling stage - we can go right from linalg ops to tiled representations.

They are still needed to get consistency IR at every step. But once the destructive update -> inplace update rewrite is done (with asymmetric ops that only exist in IREE atm and kill outwards flowing SSA values) + once we have the "memref-like-shape-only-type-that-composes-with-canonicalizations-and-transformations" then yes I agree life will be easier.

> Relates to the big issue tracking the unneeded memsets/loads that get introduced by the init tensors naive lowering today.

@benvanik yes and you may see what it looks like in [this test](https://github.com/google/iree/pull/4081/files#diff-5734b71fce2b5c8236e4e69dfc473d51a1eff072856efba5e76a780d74070245R195), see the use of `init(%11 : tensor<1x1xf32>)`: this is designed to solve this issue. Once Mahesh's stuff lands we will be able to generate many such fusions. 

> tile + fuse works with init tensors (and works without it too. See patch https://reviews.llvm.org/D93077 ). The init_tensors can be eliminated at bufferization time, but as I said above this seems to be an undue burden on the bufferization. 

Fast forward a few weeks, "true init_tensors" get killed by destructive update -> inplace update rewrite and "shape only output tensors" are just a type bitswitch away. In both cases bufferization does nothing fancy anymore, so I wouldn't overanchor on that.

> But the main issue is that now all ops need init tensor which seems too heavy weight when the values of the tensor are not really needed.

Yes and that is the switch I could not get to flip in my mind. Not so much because of tensors but because of all the unnecessary information that can be derived in 95%+ of the cases from a linalg.generic.  The remaining 5% of the cases + OpInterface differences between tensor and buffer land make me pull the brake: linalg.generic is going quite more verbose (i.e. fused pointwise with 10 outputs will have 10 extra arguments) but that is the price for consistency and explicit is better than implicit.

> Between Method 2 and Method 3, I guess we have some disagreement on what is the "cleaner" solution. Lets see whats the consensus is. We can make either work

As was pointed before, they are essentially the same from a Linalg StructuredOpsInterface consistency perspective + canonicalization + transformations (which is the only thing I have come to care about first-order). Then going from Method 2 to Method 3 is essentially the `tensor<xxx>` -> `tensor<xxx, shape_only>` bitswitch which can take the best type form in agreement with folks looking at the shape dialect (once enough interfaces are implemented to "just work" from the perspective of the first-order).

**Comment by benvanik** [comment]
I saw your discourse thread and it scared me :P Too much talk about memrefs and assumptions about allocation (and what is/is not easy there). Maybe this gets to your point about trying to make linalg.generic be everything for all uses when in reality the differences between a compiler stack that is shape-aware and one that is not are pretty big.

`tensor<7x?x24x?xt, shape_only=1>` is terrifying to me. Why not just have a real type? In IREE (and any system trying to actually understand shapes) something being a tensor type has real load-bearing meaning independent of when something is a shape type. I'm still not sure why "if you want a shape, *take a freaking shape*" seems to be so scary to the people involved here and instead there's attempts to keep shoehorning shapes into things in ways that are unnatural :(

**Comment by benvanik** [comment]
(also, !shape.shape is not useful - but look at !shapex.ranked_shape - it contains all the information of `tensor<?xt, shape_only=1>` - only with a type that plays well with the type system - having that upstreamed into the shape dialect is much cleaner than forking the type on an optional attr that everything touching TensorType will need to handle)

**Comment by benvanik** [comment]
TLDR of meeting: shape_only is a workaround for ShapedType closed type system and can be resolved (as something like shapex.ranked_shape based on some common interface is upstreamed), init tensors can be removed when not needed during tile+fuse and won't survive across dispatch region boundaries (my major concern), and tensor<shape_only> can be converted to/from shapex.ranked_shape when needed.

**Comment by MaheshRavishankar** [comment]
This has been addressed with Linalg ops for now by all ops on tensors having an `outs` operands list that can be used to get the shape only. Closing for now
